src/impls.rs

```diff
@@ -113,200 +90,122 @@ impl_metadata_for_tuple!(A, B, C, D, E, F, G, H);
 	}
 }
 
-impl<K, V> HasTypeDef for BTreeMap<K, V>
+impl<K, V> TypeInfo for BTreeMap<K, V>
 where
 	K: Metadata + 'static,
 	V: Metadata + 'static,
 {
-	fn type_def() -> TypeDef {
-		TypeDefStruct::new(vec![NamedField::new("elems", MetaType::new::<[(K, V)]>())]).into()
-	}
-}
-
-impl<T> HasTypeId for Box<T>
-where
-	T: HasTypeId + ?Sized,
-{
-	fn type_id() -> TypeId {
-		T::type_id()
+	fn type_info() -> Type {
+		TypeComposite::new("BTreeMap", Namespace::prelude())
+			.type_params(tuple_meta_type![(K, V)])
+			.fields(Fields::named().field_of::<[(K, V)]>("elems"))
+			.into()
 	}
 }
 
-impl<T> HasTypeDef for Box<T>
+impl<T> TypeInfo for Box<T>
 where
-	T: Metadata + ?Sized,
+	T: TypeInfo + ?Sized,
 {
-	fn type_def() -> TypeDef {
-		T::type_def()
+	fn type_info() -> Type {
+		T::type_info()
 	}
 }
 
-impl<T> HasTypeId for &T
+impl<T> TypeInfo for &T
 where
-	T: HasTypeId + ?Sized,
+	T: TypeInfo + ?Sized,
 {
-	fn type_id() -> TypeId {
-		T::type_id()
+	fn type_info() -> Type {
+		T::type_info()
 	}
 }
 
-impl<T> HasTypeDef for &T
+impl<T> TypeInfo for &mut T
 where
-	T: Metadata + ?Sized,
+	T: TypeInfo + ?Sized,
 {
-	fn type_def() -> TypeDef {
-		T::type_def()
+	fn type_info() -> Type {
+		T::type_info()
 	}
 }
 
-impl<T> HasTypeId for &mut T
-where
-	T: HasTypeId + ?Sized,
-{
-	fn type_id() -> TypeId {
-		T::type_id()
-	}
-}
-
-impl<T> HasTypeDef for &mut T
-where
-	T: Metadata + ?Sized,
-{
-	fn type_def() -> TypeDef {
-		T::type_def()
-	}
-}
-
-impl<T> HasTypeId for [T]
+impl<T> TypeInfo for [T]
 where
 	T: Metadata + 'static,
 {
-	fn type_id() -> TypeId {
-		TypeIdSlice::of::<T>().into()
-	}
-}
-
-impl<T> HasTypeDef for [T]
-where
-	T: Metadata,
-{
-	fn type_def() -> TypeDef {
-		TypeDef::builtin()
-	}
-}
-
-impl HasTypeId for str {
-	fn type_id() -> TypeId {
-		TypeIdPrimitive::Str.into()
+	fn type_info() -> Type {
+		TypeSlice::of::<T>().into()
 	}
 }
 
-impl HasTypeDef for str {
-	fn type_def() -> TypeDef {
-		TypeDef::builtin()
+impl TypeInfo for str {
+	fn type_info() -> Type {
+		TypePrimitive::Str.into()
 	}
 }
 
-impl HasTypeId for String {
-	fn type_id() -> TypeId {
-		<str>::type_id()
-	}
-}
-
-impl HasTypeDef for String {
-	fn type_def() -> TypeDef {
-		TypeDefStruct::new(vec![NamedField::new("vec", MetaType::new::<Vec<u8>>())]).into()
-	}
-}
-
-impl<T> HasTypeId for PhantomData<T>
-where
-	T: Metadata + ?Sized,
-{
-	fn type_id() -> TypeId {
-		TypeIdCustom::new("PhantomData", Namespace::prelude(), vec![T::meta_type()]).into()
+impl TypeInfo for String {
+	fn type_info() -> Type {
+		TypeComposite::new("String", Namespace::prelude())
+			.fields(Fields::named().field_of::<Vec<u8>>("vec"))
+			.into()
 	}
 }
 
-impl<T> HasTypeDef for PhantomData<T>
+impl<T> TypeInfo for PhantomData<T>
 where
 	T: Metadata + ?Sized,
 {
-	fn type_def() -> TypeDef {
-		TypeDefTupleStruct::new(vec![]).into()
+	fn type_info() -> Type {
+		TypeComposite::new("PhantomData", Namespace::prelude())
+			.type_params(vec![T::meta_type()])
+			.unit()
+			.into()
```

Comment:
I am actually not sure if we want to have a `TypeInfo` for types like `PhantomData` since they do not have an effect on the SCALE encoding. So instead we could try to extend the macros with an `#[type_info(ignore)]` attribute to flag those fields to be non-relevant. Or we could support the no-op (empty, never, whatever) type info variant. I am undecided what to do best. What do you think?

Reply:
I thought against about this and I really think we should get rid of types like `PhantomData` from the serialization since its job is to only serialize what is important for decoding SCALE information. `PhantomData` is not important for this job.